File: bible/setup_db.js
#!/usr/bin/env node
/**
 * setup_db.js
 * ───────────
 * Run once to create all MongoDB collections with:
 *   - JSON Schema validators (enforced by MongoDB at write time)
 *   - All indexes
 *
 * Usage:
 *   node setup_db.js
 *   node setup_db.js --uri mongodb://user:pass@host:27017 --db bible
 *
 * Requirements:
 *   npm install mongodb
 */

'use strict';

const { MongoClient } = require('mongodb');
const fs   = require('fs');
const path = require('path');

// ── CLI args ──────────────────────────────────────────────────────────────────
const args = process.argv.slice(2).reduce((acc, val, i, arr) => {
  if (val.startsWith('--')) acc[val.slice(2)] = arr[i + 1];
  return acc;
}, {});

const URI     = args.uri || 'mongodb://localhost:27017';
const DB_NAME = args.db  || 'bible';

// ── Schema files ──────────────────────────────────────────────────────────────
const SCHEMA_DIR = path.join(__dirname, 'jsonschema');

const COLLECTION_SCHEMAS = {
  editions:        'edition.schema.json',
  books:           'book.schema.json',
  chapters:        'chapter.schema.json',
  lexicon:         'lexicon.schema.json',
  verseAlignments: 'verse-alignment.schema.json',
  wordAlignments:  'word-alignment.schema.json',
  annotations:     'annotation.schema.json',
  readingProgress: 'reading-progress.schema.json',
};

// ── Helpers ───────────────────────────────────────────────────────────────────
const log = {
  info:  (msg) => console.log(`INFO   ${msg}`),
  ok:    (msg) => console.log(`✅     ${msg}`),
  warn:  (msg) => console.warn(`WARN   ${msg}`),
  error: (msg) => console.error(`ERROR  ${msg}`),
};

function loadSchema(filename) {
  const filepath = path.join(SCHEMA_DIR, filename);
  if (!fs.existsSync(filepath)) {
    log.warn(`Schema file not found: ${filepath} — skipping validator`);
    return null;
  }
  return JSON.parse(fs.readFileSync(filepath, 'utf-8'));
}

// ── Collection creation ───────────────────────────────────────────────────────
async function createCollection(db, name, schema) {
  const validator       = schema ? { $jsonSchema: schema } : {};
  const validationLevel  = 'moderate';   // enforce on new docs, warn on existing
  const validationAction = 'error';      // reject invalid inserts/updates

  const existingCollections = await db
    .listCollections({ name })
    .toArray();

  if (existingCollections.length === 0) {
    // Collection does not exist — create it
    await db.createCollection(name, {
      validator,
      validationLevel,
      validationAction,
    });
    log.info(`Created collection:   ${name}`);
  } else {
    // Collection exists — update validator only
    await db.command({
      collMod: name,
      validator,
      validationLevel,
      validationAction,
    });
    log.info(`Updated validator:    ${name}`);
  }
}

// ── Index definitions ─────────────────────────────────────────────────────────
async function createIndexes(db) {
  log.info('Creating indexes …');

  // editions
  await db.collection('editions').createIndexes([
    { key: { abbreviation: 1 }, unique: true },
    { key: { language: 1 } },
  ]);

  // books
  await db.collection('books').createIndexes([
    { key: { editionId: 1, order: 1 } },
    { key: { editionId: 1, testament: 1, order: 1 } },
    { key: { usfmCode: 1, editionId: 1 } },
  ]);

  // chapters
  await db.collection('chapters').createIndexes([
    { key: { editionId: 1, bookId: 1, order: 1 }, unique: true },
    { key: { canonicalRef: 1 } },
    { key: { 'verses.canonicalRef': 1 } },
    { key: { 'verses.strongsRefs': 1 } },
    { key: { 'verses.crossReferences': 1 } },
    { key: { plainText: 'text', 'verses.plainText': 'text' } },
  ]);

  // lexicon
  await db.collection('lexicon').createIndexes([
    { key: { strongsNumber: 1 }, unique: true },
    { key: { language: 1 } },
    { key: { lemma: 'text', definition: 'text' } },
  ]);

  // verseAlignments
  await db.collection('verseAlignments').createIndexes([
    { key: { canonicalRef: 1 }, unique: true },
  ]);

  // wordAlignments
  await db.collection('wordAlignments').createIndexes([
    { key: { canonicalRef: 1 }, unique: true },
    { key: { 'alignments.strongsRef': 1 } },
  ]);

  // annotations
  await db.collection('annotations').createIndexes([
    { key: { userId: 1, createdAt: 1 } },
    { key: { userId: 1, canonicalRef: 1 } },
    { key: { userId: 1, type: 1 } },
    { key: { userId: 1, tags: 1 } },
    { key: { groupId: 1, canonicalRef: 1 } },
    { key: { linkedCanonicalRef: 1 } },
    { key: { 'note.plainText': 'text' } },
  ]);

  // readingProgress
  await db.collection('readingProgress').createIndexes([
    { key: { userId: 1, editionId: 1 }, unique: true },
    { key: { userId: 1, 'lastRead.readAt': -1 } },
  ]);

  log.info('All indexes created.');
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function setup() {
  log.info(`Connecting to ${URI} …`);
  const client = new MongoClient(URI, { serverSelectionTimeoutMS: 5_000 });

  try {
    await client.connect();
    await client.db('admin').command({ ping: 1 });
    log.info(`Connected. Setting up database: ${DB_NAME}`);

    const db = client.db(DB_NAME);

    // 1. Create collections with validators
    log.info('Creating collections with validators …');
    for (const [collectionName, schemaFile] of Object.entries(COLLECTION_SCHEMAS)) {
      const schema = loadSchema(schemaFile);
      await createCollection(db, collectionName, schema);
    }

    // 2. Create indexes
    await createIndexes(db);

    // 3. Summary
    console.log('');
    log.ok(`Database '${DB_NAME}' is ready.`);
    log.ok(`Collections: ${Object.keys(COLLECTION_SCHEMAS).join(', ')}`);

  } catch (err) {
    log.error(`Setup failed: ${err.message}`);
    process.exit(1);
  } finally {
    await client.close();
  }
}

setup();

File: bible/setup_db.py
"""
setup_db.py
───────────
Run once to create all MongoDB collections with:
  - JSON Schema validators (enforced by MongoDB at write time)
  - All indexes

Usage:
    python setup_db.py
    python setup_db.py --uri mongodb://user:pass@host:27017 --db bible
"""

import argparse
import json
import logging
import sys
from pathlib import Path

from pymongo import MongoClient
from pymongo.errors import CollectionInvalid

logging.basicConfig(level=logging.INFO, format="%(levelname)s  %(message)s")
log = logging.getLogger(__name__)

# ── JSON Schema files (relative to this script) ───────────────────────────────
SCHEMA_DIR = Path(__file__).parent / "jsonschema"

COLLECTION_SCHEMAS = {
    "editions":        "edition.schema.json",
    "books":           "book.schema.json",
    "chapters":        "chapter.schema.json",
    "lexicon":         "lexicon.schema.json",
    "verseAlignments": "verse-alignment.schema.json",
    "wordAlignments":  "word-alignment.schema.json",
    "annotations":     "annotation.schema.json",
    "readingProgress": "reading-progress.schema.json",
}


def load_schema(filename: str) -> dict:
    path = SCHEMA_DIR / filename
    if not path.exists():
        log.warning("Schema file not found: %s — skipping validator", path)
        return {}
    return json.loads(path.read_text(encoding="utf-8"))


def create_collection(db, name: str, schema: dict) -> None:
    """Create a collection with an optional JSON Schema validator."""
    validator = {"$jsonSchema": schema} if schema else {}

    try:
        db.create_collection(
            name,
            validator=validator,
            validationLevel="moderate",   # warn on existing docs, enforce on new ones
            validationAction="error",     # reject invalid inserts/updates
        )
        log.info("  Created collection: %s", name)

    except CollectionInvalid:
        # Collection already exists — update its validator instead
        cmd = {"collMod": name, "validationLevel": "moderate", "validationAction": "error"}
        if validator:
            cmd["validator"] = validator
        db.command(cmd)
        log.info("  Updated validator:  %s", name)


def create_indexes(db) -> None:
    from pymongo import ASCENDING, TEXT

    log.info("Creating indexes …")

    db.editions.create_index("abbreviation", unique=True)
    db.editions.create_index("language")

    db.books.create_index([("editionId", ASCENDING), ("order", ASCENDING)])
    db.books.create_index([("editionId", ASCENDING), ("testament", ASCENDING), ("order", ASCENDING)])
    db.books.create_index([("usfmCode", ASCENDING), ("editionId", ASCENDING)])

    db.chapters.create_index(
        [("editionId", ASCENDING), ("bookId", ASCENDING), ("order", ASCENDING)],
        unique=True,
    )
    db.chapters.create_index("canonicalRef")
    db.chapters.create_index("verses.canonicalRef")
    db.chapters.create_index("verses.strongsRefs")
    db.chapters.create_index("verses.crossReferences")
    db.chapters.create_index([("plainText", TEXT), ("verses.plainText", TEXT)])

    db.lexicon.create_index("strongsNumber", unique=True)
    db.lexicon.create_index("language")
    db.lexicon.create_index([("lemma", TEXT), ("definition", TEXT)])

    db.verseAlignments.create_index("canonicalRef", unique=True)

    db.wordAlignments.create_index("canonicalRef", unique=True)
    db.wordAlignments.create_index("alignments.strongsRef")

    db.annotations.create_index([("userId", ASCENDING), ("createdAt", ASCENDING)])
    db.annotations.create_index([("userId", ASCENDING), ("canonicalRef", ASCENDING)])
    db.annotations.create_index([("userId", ASCENDING), ("type", ASCENDING)])
    db.annotations.create_index([("userId", ASCENDING), ("tags", ASCENDING)])
    db.annotations.create_index([("groupId", ASCENDING), ("canonicalRef", ASCENDING)])
    db.annotations.create_index("linkedCanonicalRef")
    db.annotations.create_index([("note.plainText", TEXT)])

    db.readingProgress.create_index(
        [("userId", ASCENDING), ("editionId", ASCENDING)], unique=True
    )

    log.info("All indexes created.")


def setup(uri: str, db_name: str) -> None:
    log.info("Connecting to %s …", uri)
    client = MongoClient(uri, serverSelectionTimeoutMS=5_000)

    # Verify connection
    client.admin.command("ping")
    log.info("Connected. Setting up database: %s", db_name)

    db = client[db_name]

    log.info("Creating collections with validators …")
    for collection_name, schema_file in COLLECTION_SCHEMAS.items():
        schema = load_schema(schema_file)
        create_collection(db, collection_name, schema)

    create_indexes(db)

    log.info("")
    log.info("✅  Database '%s' is ready.", db_name)
    log.info("    Collections: %s", ", ".join(COLLECTION_SCHEMAS.keys()))

    client.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Set up the Bible MongoDB database.")
    parser.add_argument("--uri", default="mongodb://localhost:27017", help="MongoDB connection URI")
    parser.add_argument("--db",  default="bible", help="Database name")
    args = parser.parse_args()

    try:
        setup(args.uri, args.db)
    except Exception as e:
        log.error("Setup failed: %s", e)
        sys.exit(1)
